### ISSUE TITLE
VisuCoreTransposition not up to date with RECO_transposition in version 5.1

### ISSUE THREAD

**Issue by SebastianoF** [SebastianoF/bruker2nifti#11]
VisuCoreTransposition should reflect the value of RECO_transposition. In some cases in version 5.1 this does not happen. This Bruker ParaVision feature, can cause problems in the re-orientation. (XY, switched with YX when RECO_transposition > 0). Some system to find the RECO_transposition from the visu_pars file alone need to be found.
Thanks to @neurolabusc for pointing this out.

**Comment by gdevenyi** [comment]
Indeed, in the PV5.1 data I have ``VisuCoreTransposition`` doesn't exist in ``visu_pairs``

I have lots of data and access to a PV5.1 manual and console, happy to share anything.

**Comment by neurolabusc** [comment]
Note images from @gdevenyi are already included with [bruker2nifti_qa](https://gitlab.com/naveau/bruker2nifti_qa). I think a large dataset should allow a dedicated developer to decipher how the formats have changed over time and a stable method to extract spatial coordinates regardless of the vintage of the data. As a professional vendor, it is unfortunate that Bruker did not document this. If anyone wants to tackle this, they could generate a pull request to support this.

**Comment by gdevenyi** [comment]
I now have PV5 and PV6.1 on the same computer attached to the same MRI. I can conceivably do some head-to-head comparisons of the same sequences collected.

Any requests?

**Comment by SebastianoF** [comment]
Oh, well... Thanks!!
Do you think an example of a 2d and a 3d sequence with both versions for the same sample (where left and right are distinguishable) may be of some help for further testing? 

We should also involve the bruker2nifti_qa maintainers in this, who may like to host the example with the current dataset (I will send Mikaël Naveau an email shortly, pointing at this page, if he is keen to participate).

Thanks again for the offer!

**Comment by SebastianoF** [comment]
Morning @gdevenyi,
Good news from the bruker2nifti_qa side. Please follow up on [this](https://gitlab.com/naveau/bruker2nifti_qa/issues/7#note_170842936) thread. @naveau has good insight into possible sequences for comparisons.
Thanks!

**Comment by neurolabusc** [comment]
@gdevenyi might be nice to have [diffusion images](https://github.com/neurolabusc/Bru2Nii/issues/23) with one set where slices are aligned orthogonal to scanner bore and another where they are rotated. Would be nice to have the NIfTI files exported by [ParaVision 360 1.1](https://github.com/neurolabusc/Bru2Nii/issues/24). I am hoping that Bruker is now able to convert their proprietary format directly to standard formats, obviating the need for 3rd party tools that rely on reverse engineering. 

@cecilyen may also have ideas.

Do the Bruker files record if the participant was scanner prone versus supine? Do they encode whether the participant is a biped or a quadruped (similar to [DICOM](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.2.html))? Since NIfTI expects coordinates in the frame of reference of the participant, this would be useful to know.

**Comment by naveau** [comment]
On PV6.0.1, the patient position is reported in these files :
```
subject: ##$SUBJECT_position=SUBJ_POS_Prone
1/acqp: ##$ACQ_patient_pos=Head_Prone
1/visu_pars: ##$VisuSubjectPosition=Head_Prone
```

The type of participant is reported here :
```
subject: ##$SUBJECT_type=Quadruped
1/visu_pars: ##$VisuSubjectType=Quadruped
1/pdata/1/visu_pars: ##$VisuSubjectType=Quadruped
```

**Comment by dvm-shlee** [comment]
@cecilyen, @neurolabusc, @SabstianoF @gdevenyi 
Hi All, I just want to let you know that I've made some progression to resolve the issue related to the patient position incorrectly. But only tested with the data in our group. So far, all tested data had been showing good orientation with each other even with an oblique FOV and different slice axis across several modalities. 

If you have any dataset that having an issue to convert with correct affine transform, I hope to ask if you can test if my python module works well with it. 

The final solution made based on the lots of trials and errors, so sorry that I couldn't answer clearly how I was able to solve it, but briefly, the issues related to difficulty in making correct orientation in PV6.0.1 were **PVM_SPackArrGradOrient** and **VisuCoreDiskSliceOrder**. The first one related to the gradient orientation, which asks as a hidden layer to control axis direction. And for **VisuCoreDiskSliceOrder** is set as **disk_reverse_slice_order** in case of the 3D acquisition, so that slice direction is flipped, so if you did not correct this, the orientation of image will be located the as the mirror along slice axis. 

As so far, my module removed these issues and available in my [GitHub](https://github.com/dvm-shlee/bruker). and If I can get your feedback, it would be really appreciated.

BTW, for your information, regarding the subject position, my code considers Supine as a Prone position in the case of the Quadruped. As you already know, because Paravision does not handle that well and showing an inverted DV image with Prone.  I tried to take into account all possible positions as much as possible, but the different cases had not been tested because I don't have sample data. If it works fine, the A-P, L-R, and S-I will be accommodated the subject orientation.

**Comment by SebastianoF** [comment]
Hi @dvm-shlee !
Thanks for your contribution. I see that you too have:

>  resolve the issue related to the incorrect patient position. But only tested with the data in our group.

and 
> if it really works well or not with other images.

Thi summarizes really well why this issue is still open. It worked on my data, but I can not be sure it works for different acquisition parametrisation. 

A possible benchmark dataset is provided by Michael Naveau with [bruker2nifti_qa](https://gitlab.com/naveau/bruker2nifti_qa/tree/master), that you can use to test your solution.

Though, an ideal pathway to unify the conversions in to a standard would be to compare the formulas of the conversions underpinning the code, rather than comparing the outcome of the code on different datasets. 

Unfortunately, I have no more access to Bruker data (I changed job... twice), so I can help on this no more. I leave it to you to test. Also, if you, or anyone is keen on taking over the maintenance of bruker2nifti, I would be happy to transfer the repo, or point at its fork, if you consider this is still worth doing after the official converter had been released.

**Comment by SebastianoF** [comment]
P.S. I added your converter to the list of open source available converters in my code documentation: https://github.com/SebastianoF/bruker2nifti/wiki/References

**Comment by neurolabusc** [comment]
@dvm-shlee I have updated my moribund [Bru2Nii](https://github.com/neurolabusc/Bru2Nii) to link to your new work. Since my site no longer has Bruker equipment and Sebastiano has moved on, the mantle for supporting this poorly documented, internally contradictory, buggy and evolving "format" (in the loosest sense of the word) has passed to you. I earnestly feel the scientific community needs someone in this role. Thank you for your service. Hopefully the direct support for NIfTI in Paravision 360 will reduce the burden on you in the long term.

**Comment by dvm-shlee** [comment]
@neurolabusc Thank you for all your advice, and really sad to hear that your site no longer maintaining Bruker equipment and Dr. Ferraris has left this field. I also feel the same that the PVdataset format is annoying to access their data a lot, hope they did not change internally again to make it hard to maintain the third-party tools. The reason I've done this recently is due to the COVID-19 stop all of my animal research while I haven't expected they added NifTi converter on their new software. But since the Paravision 360 requiring the new hardware AvanceNeo which most of the groups in the academic field are not easy to upgrade soon.

I think to maintain the third-party converter is still required for the researcher who needs to reuse their old data, needs to access FID data and acquisition to develop reconstruction tool, or need to convert multiple datasets with some automated manner (which is once of my priority). Since your group had been accumulated lots of effort on this even you are no longer have Bruker, I will try to accommodate your effort into my repo as much as possible, and will not miss granting the credit for all your prior works.

@SebastianoF Thank you for really helpful advice and for sharing your idea including some other at your email. First, I haven't even submitted this work to the paper, but I just prepared a draft because I don't want to waste my effort on just converting our internal data. And actually, your previous efforts in sharing your work with the community inspired me a lot recently. 

Sorry for you if you have any feeling that I've spooned anything from your prior works and does not put any credit of your work on my manuscript, but I really haven't take a look much the other people works before COVID-19 pandemic since I start made my first converter 2015 when our magnet quenched for my own data conversion. Frankly, I found your code around 2018 which I hoped that the issues were resolved by others but really did not check issues or your publication because just I was not familiar with the Github system at that time. So that time I decided to start updating my effort on my converter into Github to have better version control because I was struggling to maintain my code while other colleagues complained about my code in terms of stability. In addition to that, the performance of my code was relatively naive compared to this repo in terms of converting other modality than EPI. 

Recently I have realized that you have been published your work in 2017 and build a great community to initiating the collaboration. This is the main reason I first post on this thread to get your feedback related to this issue, before sharing or submit this to the public domain. (my original plan was to share this code to experimental fMRI colleagues who using Bruker scanner for making standard pipeline, so only other I've been contacted so far is BIDS community since I have been focussed on the data organization more.). 

Regarding the modality, I've tested pretty many cases using the data in our center which I had issues so far for conversion for internal testing which I cannot share to the public. But I'm sure that within-session, whatever your parameter is, the coordinate space will be the same (magnet center is image center), and I'm confident about this (but it would be also appreciated if there any dataset conflict with my code.). The one that I mentioned 'I was not sure' is the accommodation of species and position related to the parameter. It handled at the last layer of converting affine in my code [(code link)](https://github.com/dvm-shlee/bruker/blob/dd34451e0bacb5698fe0ee67d83c203ddf924eeb/brkraw/lib/utils.py#L40) by rotating the affine matrix based on the animal position. so unless Bruker changes their way to construct an affine matrix, it would be work fine in terms of preserving the orientation within the same session. If some conditions wrong, I just need to change the parameter to rotate the affine matrix. So the feedback from the other site would be greatly appreciated.

Regarding your feedback on python module design, really appreciated your feedback. I actually recently started feeling it in a similar manner, so the fMRI analysis tool is developed as an extension module to use the brkraw module as an API to access data [the link](https://github.com/dvm-shlee/bruker_fmri) as well as all other projects I have been working on, which make me spend a lot of time to reorganize my repositories but after that, the productivity was improved a lot. But really appreciate your attitude for sharing your know-how. Especially, your way to describe as Lego helped me a lot to build a clear concept on that for my future project! I'm really sad that the talented researcher like you is no longer in the same field (preclinical MRI). Again I really appreciate a lot of valuable comments and advice. Not sure I can publish this work soon to JOSS since the submission process is suspended, but if you have any comments on my manuscript will be greatly appreciated.

**Comment by dvm-shlee** [comment]
@SebastianoF Again, thank you for your advice regarding testing the QA dataset shared by @naveau. It actually reveals two issues from the dataset I haven't tested before, which is really good but still promising. Below is the result of the conversion, and I make it transparent to overlap each other within the session, and the last image for between session. And also summarize the issue of my code still need to be addressed at the end of this comment.

![QA](https://user-images.githubusercontent.com/7221078/81590222-7b330d00-9388-11ea-93e5-9e23dcfec66d.png)

The first issue was the case of using advanced slicing, which I need some additional correction of my code to dealing with multiple slice packs dataset. The second issue is I think it related to the scan number 5 and 9 in 20130412_APM_DEV_orient that image shows offset, which I think related to the rotation of FOV, so the origin point of matrix is messed up, (not sure yet this is the case since I haven't take a look deeply,) which I've been corrected for PV6.0.1 but not for PV5. 

This is really helpful dataset! Not sure you are still interested in this, but I will update here if I made any progression.

Also if there are any dataset had issues before so need to be tested, please let me know, I will try to check if those issues could be resolved with my code.

Thanks!

**Comment by dvm-shlee** [comment]
![Bruker2Nifti_QA_challenge](https://user-images.githubusercontent.com/7221078/81628088-be639f00-93cd-11ea-8219-7fc0d01e6113.png)

I've just corrected all the issues, All Bruker2Nifti_QA data converted with corrected orientation,
centered at the magnet center.
The only thing I struggled with was intensity slope correction for reconstructed DTI. So please let me know if you have any advice on that.

The reconstructed DTI (RecoID 2) of scan#  2 and 3, has 22 frames, but the VisuCoreDataSlope has 110 values, do you have any clue how to deal with this? I'm not familiar with the DTI image, so I need help if this is a crucial step need to be applied.

**Comment by SebastianoF** [comment]
> Sorry for you if you have any feeling that I've spooned anything from your prior works and does not put any credit of your work on my manuscript.

Not at all!! The code is there to be used, and I do not need credit, and I do not have any feeling at all either. :see_no_evil: 

The point was for you to not re-do anything that is already there, though sometimes re-doing things already done is the best way to learn and to improve open source code.

About the VisuCoreDataSlope question, I think I had a similar issue. Though in my dataset I have not been correcting for the slope, which was how I solved it.

I'll let you know here if something else comes up to my mind!

**Comment by SebastianoF** [comment]
Ah, yes, I could omit the slope because all the values in VisuCoreDataSlope had the same magnitude.

If this is not your case, then you can ask the question on a new issue thread (or maybe on stackoverflow). Here it may not be very visible.

**Comment by gdevenyi** [comment]
This is amazing work. Great job!

Here I offer a new phantom I have developed, and a scan on PV5.1. I have PV 6 now, and I intend to run a bunch of scans against this in PV5 and PV6, but... COVID. It'll happen eventually.


https://we.tl/t-Sgi1bHflNQ
